homeassistant/components/rfxtrx/device_action.py

```diff
@@ -0,0 +1,78 @@
+"""Provides device automations for RFXCOM RFXtrx."""
+from __future__ import annotations
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_DEVICE_ID, CONF_DOMAIN, CONF_TYPE
+from homeassistant.core import Context, HomeAssistant
+import homeassistant.helpers.config_validation as cv
+
+from . import _LOGGER, DATA_RFXOBJECT, DOMAIN
+from .helpers import async_get_device_object
+
+CONF_DATA = "data"
+CONF_SUBTYPE = "subtype"
+
+ACTION_TYPE_COMMAND = "send_command"
+ACTION_TYPE_STATUS = "send_status"
+
+ACTION_TYPES = {
+    ACTION_TYPE_COMMAND,
+    ACTION_TYPE_STATUS,
+}
+
+ACTION_SELECTION = {
+    ACTION_TYPE_COMMAND: "COMMANDS",
+    ACTION_TYPE_STATUS: "STATUS",
+}
+
+ACTION_SCHEMA = cv.DEVICE_ACTION_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_TYPE): vol.In(ACTION_TYPES),
+        vol.Required(CONF_SUBTYPE): str,
+    }
+)
+
+
+async def async_get_actions(hass: HomeAssistant, device_id: str) -> list[dict]:
+    """List device actions for RFXCOM RFXtrx devices."""
+    actions = []
+
+    device = async_get_device_object(hass, device_id)
+    if device:
+        for action_type in ACTION_TYPES:
+            if hasattr(device, action_type):
+                values = getattr(device, ACTION_SELECTION[action_type], {})
+                for value in values.values():
+                    actions.append(
+                        {
+                            CONF_DEVICE_ID: device_id,
+                            CONF_DOMAIN: DOMAIN,
+                            CONF_TYPE: action_type,
+                            CONF_SUBTYPE: value,
+                        }
+                    )
+
+    return actions
+
+
+async def async_call_action_from_config(
+    hass: HomeAssistant, config: dict, variables: dict, context: Context | None
+) -> None:
+    """Execute a device action."""
+    config = ACTION_SCHEMA(config)
+
+    rfx = hass.data[DOMAIN][DATA_RFXOBJECT]
+    device = async_get_device_object(hass, config[CONF_DEVICE_ID])
+
+    action_type = config[CONF_TYPE]
+    sub_type = config[CONF_SUBTYPE]
+    send_fun = getattr(device, action_type)
+    commands = getattr(device, ACTION_SELECTION[action_type], {})
+
+    for key, value in commands.items():
+        if value == sub_type:
+            await hass.async_add_executor_job(send_fun, rfx.transport, key)
+            break
+    else:
+        _LOGGER.error("Subtype %s not found in device commands %s", sub_type, commands)
```

Comment:
Move this check to when the automation or script is setup instead by defining a function `async_validate_action_config`. If that function is defined it will be called to validate the action instead of using the `ACTION_SCHEMA`.

Reply:
Technically it cant happen. You can not setup this invalid subtype. 

It can only happen by manual changes to yaml, or breaking changes in lib. So not sure what the benefit would be.

I would still need the check here, since i need to do the lookup. It could just raise an exception instead.

Reply:
The benefit is to give helpful error messages during automation and script setup, ideally with filename + line of the yaml file instead of a fairly anonymous log message.

You can remove this check because it's impossible for an automation or script with an invalid action to be setup.

Here's an example from deconz (although for device trigger): https://github.com/home-assistant/core/blob/abca7deadbb04230a383e40cf068a15210a2856d/homeassistant/components/deconz/device_trigger.py#L601

Reply:
I can add that. But i can't remove the check anyway. I need to do a reverse dict lookup. So the for loop need to stay anyways. There lines of else: log, break is not much of a reduction in complexity.

But i don't mind adding the validation function.

Reply:
I have added validation now.